cpp/src/arrow/compute/kernels/aggregate_mode.cc

```diff
@@ -26,95 +26,187 @@ namespace aggregate {
 
 namespace {
 
+// Count values and return value:count map
+template <typename T>
+struct ValueCounter {
+  virtual void CountOne(T value) = 0;
```

Comment:
I don't think polymorphism is a good idea for performance. Instead, you should probably use templated code.
Also, when `null_count == length`, the implementation can be trivial.

Reply:
Changed to template. Did get better performance. Thanks.